# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [2]:
import pandas as pd

In [3]:
original_data = pd.read_csv('e_commerce.csv')
original_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

# 评估数据 

在这一部分，我将对在上一部分建立的`original_data`这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

###  评估数据整齐度

In [9]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
527948    580727     23171            REGENCY TEA PLATE GREEN          1   
221014    556237     23264  SET OF 3 WOODEN SLEIGH DECORATIONS         1   
290017    562347     23360        SET 8 CANDLES VINTAGE DOILEY         1   
424352    573264     23329       DECORATIVE WICKER HEART LARGE         1   
147967    549155     22713                 CARD I LOVE LONDON          1   
511060    579471     23171            REGENCY TEA PLATE GREEN          1   
123272    546896     22844    VINTAGE CREAM DOG FOOD CONTAINER         1   
525976    580644     23148   MINIATURE ANTIQUE ROSE HOOK IVORY         6   
64983     541684     22665   RECIPE BOX BLUE SKETCHBOOK DESIGN         2   
276200    561036     21172                   PARTY METAL SIGN          1   

             InvoiceDate  UnitPrice  CustomerID         Country  
527948   12/5/2011 17:17       3.29     14096.0  United Kingdom  
221014    6/9/2011 15:34       2.50         NaN  United Kingdom  
290017    8/4/2011 12:22       1.95     13263.0  United Kingdom  
424352  10/28/2011 12:36       1.65     14415.0  United Kingdom  
147967    4/6/2011 17:08       0.83         NaN  United Kingdom  
511060  11/29/2011 15:08       3.29         NaN  United Kingdom  
123272   3/17/2011 18:24      16.63         NaN  United Kingdom  
525976   12/5/2011 13:10       0.83     16904.0  United Kingdom  
64983    1/20/2011 15:22       2.95     15498.0  United Kingdom  
276200   7/24/2011 11:54       1.45     13137.0  United Kingdom

In [5]:
original_data.set_index('InvoiceNo')

StockCode                          Description  Quantity  \
InvoiceNo                                                            
536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
536365        71053                  WHITE METAL LANTERN         6   
536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...             ...                                  ...       ...   
581587        22613          PACK OF 20 SPACEBOY NAPKINS        12   
581587        22899         CHILDREN'S APRON DOLLY GIRL          6   
581587        23254        CHILDRENS CUTLERY DOLLY GIRL          4   
581587        23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
581587        22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
InvoiceNo                                                          
536365      12/1/2010 8:26       2.55     17850.0  United Kingdom  
536365      12/1/2010 8:26       3.39     17850.0  United Kingdom  
536365      12/1/2010 8:26       2.75     17850.0  United Kingdom  
536365      12/1/2010 8:26       3.39     17850.0  United Kingdom  
536365      12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
581587     12/9/2011 12:50       0.85     12680.0          France  
581587     12/9/2011 12:50       2.10     12680.0          France  
581587     12/9/2011 12:50       4.15     12680.0          France  
581587     12/9/2011 12:50       4.15     12680.0          France  
581587     12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 7 columns]

从抽样的10行数据结构来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交
易，每列是交易相关的各个变量，因此不存在结构性问题。

###  评估数据干净程度

In [10]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


InvoiceNo StockCode                      Description  Quantity  \
541904    581587     22613      PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899     CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254    CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID Country  
541904  12/9/2011 12:50       0.85     12680.0  France  
541905  12/9/2011 12:50       2.10     12680.0  France  
541906  12/9/2011 12:50       4.15     12680.0  France  
541907  12/9/2011 12:50       4.15     12680.0  France  
541908  12/9/2011 12:50       4.95     12680.0  France

从数据结果来看，数据共有541909条观察值，而Description, Customer ID变量存在缺失值。
此外，InvoiceDate的数据类型应为日期，Customer ID的数据类型应为字符串，应当进行数据格式转换。

###  评估缺失数据

在了解Description存在缺失值后，根据条件提取出缺失观察值。

In [14]:
original_data[original_data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失Description变量值。  
从输出结果来看，这些缺失Description的交易数据，UnitPrice都为0，为了验证猜想，我们增加筛选条件，看是否存在Description变量
缺失且UnitPrice不为0的数据。


In [16]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为0条，说明缺失 Description 值的数据，同时也不具备有效的 UnitPrice 值。   
Description 表示产品名称， UnitPrice 表示产品单价，都是进行后续商品交易分析的重要变量。如果他们同时缺失/无效，  
我们认为数据无法提供有效含义，因此这些数据后续可以被删除。

CustomerID 变量同样存在缺失值，因此也根据条件提取出缺失观察值。

In [18]:
original_data[original_data['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID 表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些 CustomerID 缺失的交易数据仍然有效，
因此保留此变量为空的观察值。

######  评估重复数据

根据数据变量的含义来看，虽然InvoiceNo、StockCode 和 CustomerID 都是唯一标识符，单一次交易可能包含多件商品，因此 InvoiceNo  
可以存在重复，不同交易可以包含同件商品，因此 StockCode 可以存在重复，顾客可以进行多次交易或下单多个商品，因此CustomerID 也  
可以存在重复。  
那么针对此数据集，我们无需评估重复数据。

######  评估不一致数据

不一致数据可能存在于Country变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [23]:
original_data['Country'].value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European

从Country 变量值来看，"USA"、"United States" 均在表示美国，"United Kingdom"、"UK"、"U.K."均在表示英国，   
因此应该在清洗步骤例对这些进行统一，只保留一个指代值。


###### 评估无效或错误数据

可以通过DataFrame的describe方法，对数据统计信息进行快速了解。

In [24]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

从输出结果来看，Quantity和UnitPrice 存在负数，会对后续数值分析产生影响。  
因此，我们先筛选出 Quantity 数值为负数的观察值，进一步评估起含义。

In [26]:
original_data[original_data['Quantity'] <= 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

In [28]:
original_data[(original_data['Quantity'] <= 0) & (original_data['InvoiceNo'].str[0] != 'C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

In [31]:
original_data["Country"] = original_data["Country"].replace({("UK","U.K."): "United Kingdom"})

C:\Users\admin\anaconda3\lib\site-packages\pandas\core\array_algos\replace.py:86: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  op = lambda x: operator.eq(x, b)


TypeError: Cannot compare types 'ndarray(dtype=object)' and 'tuple'